### PR TITLE
Support explain plan for the TDS adapter

### PR DESCRIPTION
**Pull request by leandrocp**
~~That's still a draft but it's working so you can test it locally. I'm not a SQL Server expert but looks like the result set has all the information we would need for a v1 (like the screenshot on the link below), so no need to parse XML.~~

I believe it's good enough for review, see https://github.com/elixir-ecto/ecto_sql/pull/273#issuecomment-703269766.

Also see https://github.com/elixir-ecto/ecto_sql/pull/231#discussion_r433391558 for reference.

Wdyt?

### CODE REVIEW

File: lib/ecto/adapters/tds/connection.ex [hunk 1]
```diff
@@ -300,10 +300,74 @@ if Code.ensure_loaded?(Tds) do
     end
 
     @impl true
-    def explain_query(_conn, _query, _params, _opts) do
-      raise Tds.Error, "EXPLAIN is not supported by Ecto.Adapters.TDS at the moment"
+    def explain_query(conn, query, params, opts) do
+      params = prepare_params(params)
+
+      case Tds.query_multi(conn, build_explain_query(query), params, opts) do
+        {:ok, [_, %Tds.Result{} = result, _]} ->
```
Comment (leandrocp) [inline lib/ecto/adapters/tds/connection.ex:307]: Not sure if the order is deterministic, but looks like it's (still need to confirm on TDS code).

File: lib/ecto/adapters/tds/connection.ex [hunk 2]
```diff
@@ -300,10 +300,74 @@ if Code.ensure_loaded?(Tds) do
...
+        "SET STATISTICS PROFILE ON; ",
+        query,
+        "; ",
+        "SET STATISTICS XML OFF; ",
+        "SET STATISTICS PROFILE OFF;"
+      ]
+      |> IO.iodata_to_binary()
+    end
+
+    defp format_result_as_table(columns, rows) do
```
Comment (leandrocp) [inline lib/ecto/adapters/tds/connection.ex:327]: Those function are exactly the same as used on myxql adapter, should I move those to `Ecto.Adapters.SQL` and put a `@doc false`?

Comment (josevalim) [reply lib/ecto/adapters/tds/connection.ex:327]: I think we can introduce a function called `format_table` in there and make it generally available. :)

**Comment by josevalim** [comment]
Looks great, awesome job! ❤️

**Comment by leandrocp** [comment]
@josevalim that's how a TDS explain looks like:

```
+------+----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+--------+--------+----------------------+----------------------+------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------+-----------------------+-----------------------+------------+-----------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+----------+----------+--------------------+
| Rows | Executes | StmtText                                                                                                                                                                                                                                                                                                                   | StmtId | NodeId | Parent | PhysicalOp           | LogicalOp            | Argument                                                                                                                           | DefinedValues                                                                                                                                                                                                                                                                            | EstimateRows | EstimateIO            | EstimateCPU           | AvgRowSize | TotalSubtreeCost      | OutputList                                                                                                                                                                                                                                                                               | Warnings | Type     | Parallel | EstimateExecutions |
+------+----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+--------+--------+----------------------+----------------------+------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------+-----------------------+-----------------------+------------+-----------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+----------+----------+--------------------+
|    0 |        1 | SELECT TOP(1) p0.[id], p0.[counter], p0.[title], p0.[blob], p0.[public], p0.[cost], p0.[visits], p0.[wrapped_visits], p0.[intensity], p0.[bid], p0.[uuid], p0.[meta], p0.[links], p0.[intensities], p0.[posted], p0.[author_id], p0.[inserted_at], p0.[updated_at] FROM [posts] AS p0 WHERE (p0.[title] = N'explain_test') |      1 |      1 |      0 | NULL                 | NULL                 | NULL                                                                                                                               | NULL                                                                                                                                                                                                                                                                                     |          1.0 | NULL                  | NULL                  | NULL       |  0.003283380065113306 | NULL                                                                                                                                                                                                                                                                                     | NULL     | SELECT   | false    | NULL               |
|    0 |        1 |   |--Top(TOP EXPRESSION:((1)))                                                                                                                                                                                                                                                                                             |      1 |      2 |      1 | Top                  | Top                  | TOP EXPRESSION:((1))                                                                                                               | NULL                                                                                                                                                                                                                                                                                     |          1.0 |                   0.0 | 1.0000000116860974e-7 |      16306 |  0.003283380065113306 | [p0].[id], [p0].[title], [p0].[counter], [p0].[blob], [p0].[bid], [p0].[uuid], [p0].[meta], [p0].[links], [p0].[intensities], [p0].[public], [p0].[cost], [p0].[visits], [p0].[wrapped_visits], [p0].[intensity], [p0].[author_id], [p0].[posted], [p0].[inserted_at], [p0].[updated_at] | NULL     | PLAN_ROW | false    |                1.0 |
|    0 |        1 |        |--Clustered Index Scan(OBJECT:([ecto_test].[dbo].[posts].[posts_pkey] AS [p0]), WHERE:([ecto_test].[dbo].[posts].[title] as [p0].[title]=N'explain_test'))                                                                                                                                                         |      1 |      3 |      2 | Clustered Index Scan | Clustered Index Scan | OBJECT:([ecto_test].[dbo].[posts].[posts_pkey] AS [p0]), WHERE:([ecto_test].[dbo].[posts].[title] as [p0].[title]=N'explain_test') | [p0].[id], [p0].[title], [p0].[counter], [p0].[blob], [p0].[bid], [p0].[uuid], [p0].[meta], [p0].[links], [p0].[intensities], [p0].[public], [p0].[cost], [p0].[visits], [p0].[wrapped_visits], [p0].[intensity], [p0].[author_id], [p0].[posted], [p0].[inserted_at], [p0].[updated_at] |          1.0 | 0.0031250000465661287 | 1.5809999604243785e-4 |      16306 | 0.0032830999698489904 | [p0].[id], [p0].[title], [p0].[counter], [p0].[blob], [p0].[bid], [p0].[uuid], [p0].[meta], [p0].[links], [p0].[intensities], [p0].[public], [p0].[cost], [p0].[visits], [p0].[wrapped_visits], [p0].[intensity], [p0].[author_id], [p0].[posted], [p0].[inserted_at], [p0].[updated_at] | NULL     | PLAN_ROW | false    |                1.0 |
+------+----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+--------+--------+----------------------+----------------------+------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------+-----------------------+-----------------------+------------+-----------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+----------+----------+--------------------+
```

That will look ugly when printed to a console that wraps the lines, but you can copy&paste to an editor, save to a file, or configure the console if that's possible. Anyway, I suppose that's good enough for the version. Wdyt?

**Comment by josevalim** [comment]
Awesome job! :heart:

**Pull request merged**